In [1]:
import utils

# Préparation des données

In [2]:
# Players lists
players_info_list = utils.get_dataframes_csv_file('../scraping/csv/players_infos')
players_stats_list = utils.get_dataframes_csv_file('../scraping/csv/players_stats')

# Teams lists
schedule_df_list = utils.get_dataframes_csv_file('../scraping/csv/schedule' )
standing_df_list = utils.get_dataframes_csv_file('../scraping/csv/standing')
team_stats_df_list = utils.get_dataframes_csv_file('../scraping/csv/team_stats')
team_stats_opponent_df_list = utils.get_dataframes_csv_file('../scraping/csv/team_stats_opponent')
team_stats_differential_df_list = utils.get_dataframes_csv_file('../scraping/csv/team_stats_differential')

In [3]:
# Trier les données en fonction des années 
players_info_list.sort(key=lambda df: df['Title'][0].split('-')[0][-4:])
players_stats_list.sort(key=lambda df: df['Title'][0].split('-')[0][-4:])

In [4]:
# Diviser les données des stats des joueurs en fonctions du type de saison 
players_stats_list_summer_league = []
players_stats_list_playoffs = []
players_stats_list_regular_season = []
players_stats_list_preseason = []

for df in players_stats_list:
    year = df['Title'].iloc[0].split(' ')[1]

    if year == '2023-2024':
        continue

    season_type = df['Title'].iloc[0].split(' ')[2]
    
    # Ajouter les données au dictionnaire approprié
    if season_type == 'Playoffs':
        players_stats_list_playoffs.append(df)
    elif season_type == 'Regular_Season':
        players_stats_list_regular_season.append(df)
    elif season_type == 'Preseason':
        players_stats_list_preseason.append(df)

In [5]:
# Diviser les données des matchs des equipes en fonctions du type de saison 
schedule_df_list_list_summer_league = []
schedule_df_list_list_playin = []
schedule_df_list_list_regular_season = []
schedule_df_list_list_preseason = []

for df in schedule_df_list:
    season_type = df['Title'].iloc[0].split(' ')[-3]
    
    # Ajouter les données au dictionnaire approprié
    if season_type == 'Summer_League':
        schedule_df_list_list_summer_league.append(df)
    elif season_type == 'Play-In':
        schedule_df_list_list_playin.append(df)
    elif season_type == 'Season':
        schedule_df_list_list_regular_season.append(df)
    elif season_type == 'Preseason':
        schedule_df_list_list_preseason.append(df)

In [7]:
# Selectionner depuis l'année 2000
players_info_list_2000 = players_info_list[54::]
players_stats_list_2000 = players_stats_list[108::]

# Selectionner depuis l'année 1990
players_info_list_1990 = players_info_list[44::]
players_stats_list_1990 = players_stats_list[88::]

# Selectionner depuis l'année 1980
players_info_list_1980 = players_info_list[34::]
players_stats_list_1980 = players_stats_list[68::]

# Selectionner depuis l'année 1970
players_info_list_1970 = players_info_list[24::]
players_stats_list_1970 = players_stats_list[48::]

# Selectionner depuis l'année 1960
players_info_list_1960 = players_info_list[14::]
players_stats_list_1960 = players_stats_list[28::]

# Selectionner depuis l'année 1950
players_info_list_1950 = players_info_list[4::]
players_stats_list_1950 = players_stats_list[8::]

# Analyse quantitif

In [8]:
print("nombre de saisons pour les statistiques des joueurs : ", len(players_info_list))

print("nombre de données de \'summer league season\' pour les statistiques des joueurs : ", len(players_stats_list_summer_league))
print("nombre de données de \'pre season\' pour les statistiques des joueurs : ", len(players_stats_list_preseason))
print("nombre de données de \'regular season\' pour les statistiques des joueurs : ", len(players_stats_list_regular_season))
print("nombre de données de \'playoff\' pour les statistiques des joueurs : ", len(players_stats_list_playoffs))

nombre de saisons pour les statistiques des joueurs :  78
nombre de données de 'summer league season' pour les statistiques des joueurs :  0
nombre de données de 'pre season' pour les statistiques des joueurs :  13
nombre de données de 'regular season' pour les statistiques des joueurs :  77
nombre de données de 'playoff' pour les statistiques des joueurs :  77


# Vérification des données

In [9]:
from collections import defaultdict

def get_null_counts(df_list):
    """
    Calcule le nombre total d'éléments nuls pour chaque colonne dans une liste de DataFrames.
    """
    null_counts = defaultdict(int)

    for col in df_list[0].columns:
        null_counts[col] = 0

    for df in df_list:
        # Compter les valeurs nulles pour chaque colonne dans ce DataFrame
        null_counts_df = df.isnull().sum().to_dict()
        
        # Ajouter ces comptes au dictionnaire total
        for col, count in null_counts_df.items():
            null_counts[col] += count

    return null_counts

def get_empty_columns_counts(df_list):
    """
    Calcule les comptes de valeurs vides, des '-' pour les chaînes de caractères.
    """
    empty_counts = defaultdict(int)

    for col in df_list[0].columns:
        empty_counts[col] = 0

    for df in df_list:
        for col in df.columns:
            # Vérifier le type de la colonne
            if df[col].dtype == object:  # Colonne de type chaîne de caractères
                if (df[col]== '').all():
                    empty_counts[col] += 1
                elif (df[col]== '-').all():
                    empty_counts[col] += 1
            else:  # Colonne numérique
                if (df[col] == 0).all():
                    empty_counts[col] += 1
    return empty_counts

### Vérification pour les données des informations de joueurs (Players infos)

In [10]:
total_stats = 0
for df in players_info_list:
    total_stats += len(df)
print("Nombre de statistique de joueurs", total_stats)


Nombre de statistique de joueurs 24561


In [11]:
print('\nNombre de valeurs null')
for col,nb in get_null_counts(players_info_list).items():
    print(col, ':', nb)


Nombre de valeurs null
Player : 0
Pos : 7
HT : 0
WT : 0
Age : 0
Teams : 0
GP : 0
YOS : 0
Pre-Draft Team : 443
Draft Status : 0
Nationality : 0
Title : 0
# : 0


<span style="color:green">
Pre-Draft Team nulls : 443 / 24561 valeurs nulls ---> ok (pas significatif)

Généralement pas de valeurs nulls --> ok
</span>

In [12]:
print('\nNombre de columns == 0 ou \'\'')
for col,nb in get_empty_columns_counts(players_info_list).items():
    print(col, ' : ', nb)


Nombre de columns == 0 ou ''
Player  :  0
Pos  :  0
HT  :  0
WT  :  0
Age  :  0
Teams  :  0
GP  :  0
YOS  :  1
Pre-Draft Team  :  0
Draft Status  :  0
Nationality  :  0
Title  :  0


<span style="color:green">
Génralement pas de valeurs vide ---> ok


La colonne YOS (Années d'expérience du joueur dans la NBA avant cette saison) est == 0 pour tout les joueurs de la premiere saison de la NBA (la saison 1946-1947) car il n'yavait pas de saison avant 
</span>

<span style="color:green">
Les données des informations de joueurs sont complets
</span>

### Vérification pour les données des statistiques des joueurs (Players stats)

In [13]:
total_stats = 0
for df in players_stats_list:
    total_stats += len(df)
print("Nombre de statistique de joueurs", total_stats)

Nombre de statistique de joueurs 49562


In [14]:
print('\nNombre de valeurs null')
for col,nb in get_null_counts(players_stats_list).items():
    print(col, ':', nb)


Nombre de valeurs null
Player : 0
Team : 2
GP : 0
MIN : 0
PTS : 0
FGM : 0
FGA : 0
FG% : 0
3PM : 0
3PA : 0
3P% : 0
FTM : 0
FTA : 0
FT% : 0
ORB : 0
DRB : 0
REB : 0
AST : 0
STL : 0
BLK : 0
TOV : 0
PF : 0
Title : 0


<span style="color:green">
Génralement pas de valeurs vide ---> ok
</span>

In [15]:
print(len(players_stats_list))

187


In [16]:
print('\nNombre de valeurs == 0 ou \'\'')
for col,nb in get_empty_columns_counts(players_stats_list).items():
    print(col, ' : ', nb)


Nombre de valeurs == 0 ou ''
Player  :  0
Team  :  0
GP  :  0
MIN  :  10
PTS  :  0
FGM  :  0
FGA  :  0
FG%  :  0
3PM  :  66
3PA  :  66
3P%  :  66
FTM  :  0
FTA  :  0
FT%  :  0
ORB  :  54
DRB  :  54
REB  :  8
AST  :  0
STL  :  54
BLK  :  54
TOV  :  62
PF  :  0
Title  :  0


<span style="color:red">
Les statistiques 3PM,3PA,3P%,ORB,DRB,STL,BLK,TOV ne sont pas contabilisé entre 54 et 66 fois sur 187---> c'est très élevé : on constate que ces données manques et cela causer problèmes

La statistique MIN n'est pas contabilisé 10 fois ---> cela pourrais aussi causé problème

</span>

On verra dans ce qui suit d'ou viennent les données manquantes en essayant les données a partir d'années différentes

In [17]:
print('\nNombre de valeurs == 0 ou \'\'')
for col,nb in get_empty_columns_counts(players_stats_list_1950).items():
    print(col, ' : ', nb)


Nombre de valeurs == 0 ou ''
Player  :  0
Team  :  0
GP  :  0
MIN  :  2
PTS  :  0
FGM  :  0
FGA  :  0
FG%  :  0
3PM  :  58
3PA  :  58
3P%  :  58
FTM  :  0
FTA  :  0
FT%  :  0
ORB  :  46
DRB  :  46
REB  :  0
AST  :  0
STL  :  46
BLK  :  46
TOV  :  54
PF  :  0
Title  :  0


In [18]:
print('\nNombre de valeurs == 0 ou \'\'')
for col,nb in get_empty_columns_counts(players_stats_list_1960).items():
    print(col, ' : ', nb)


Nombre de valeurs == 0 ou ''
Player  :  0
Team  :  0
GP  :  0
MIN  :  0
PTS  :  0
FGM  :  0
FGA  :  0
FG%  :  0
3PM  :  38
3PA  :  38
3P%  :  38
FTM  :  0
FTA  :  0
FT%  :  0
ORB  :  26
DRB  :  26
REB  :  0
AST  :  0
STL  :  26
BLK  :  26
TOV  :  34
PF  :  0
Title  :  0


In [19]:
print('\nNombre de valeurs == 0 ou \'\'')
for col,nb in get_empty_columns_counts(players_stats_list_1970).items():
    print(col, ' : ', nb)


Nombre de valeurs == 0 ou ''
Player  :  0
Team  :  0
GP  :  0
MIN  :  0
PTS  :  0
FGM  :  0
FGA  :  0
FG%  :  0
3PM  :  18
3PA  :  18
3P%  :  18
FTM  :  0
FTA  :  0
FT%  :  0
ORB  :  6
DRB  :  6
REB  :  0
AST  :  0
STL  :  6
BLK  :  6
TOV  :  14
PF  :  0
Title  :  0


In [20]:
print('\nNombre de valeurs == 0 ou \'\'')
for col,nb in get_empty_columns_counts(players_stats_list_1980).items():
    print(col, ' : ', nb)


Nombre de valeurs == 0 ou ''
Player  :  0
Team  :  0
GP  :  0
MIN  :  0
PTS  :  0
FGM  :  0
FGA  :  0
FG%  :  0
3PM  :  0
3PA  :  0
3P%  :  0
FTM  :  0
FTA  :  0
FT%  :  0
ORB  :  0
DRB  :  0
REB  :  0
AST  :  0
STL  :  0
BLK  :  0
TOV  :  0
PF  :  0
Title  :  0


<span style="color:green">

On remarque que a partir des années 80s les données des statistiques sont completes

Avant les années 80s les données de 3PM,3PA,3P%,ORB,DRB,STL,BLK,TOV  ne sont pas disponible

</span>

### Vérification pour les données des matchs des equipes (Schedules)

In [21]:
schedule_df_list = utils.get_dataframes_csv_file('../scraping/csv/schedule' )
standing_df_list = utils.get_dataframes_csv_file('../scraping/csv/standing')
team_stats_df_list = utils.get_dataframes_csv_file('../scraping/csv/team_stats')
team_stats_opponent_df_list = utils.get_dataframes_csv_file('../scraping/csv/team_stats_opponent')
team_stats_differential_df_list = utils.get_dataframes_csv_file('../scraping/csv/team_stats_differential')

In [22]:
total_lines = 0
for df in schedule_df_list:
    total_lines += len(df)
print("Nombre de matchs totales :", total_lines)

Nombre de matchs totales : 60380


In [23]:
print('\nNombre de valeurs null')
for col,nb in get_null_counts(schedule_df_list).items():
    print(col, ':', nb)


Nombre de valeurs null
DATE : 0
OPPONENT : 0
RESULT : 0
W-L : 0
Hi Points : 2175
Hi Rebounds : 2175
Hi Assists : 2175
Home_Away : 0
Title : 0


<span style="color:red">
Les columns Hi Points, Hi Rebounds, Hi Assists ont 2175 sur 60380 valeurs nulls 
</span>



<span style="color:green">
Pas de valeurs vides pour les autres colonnes 
</span>

In [24]:
print('\nNombre de valeurs null')
for col,nb in get_null_counts(schedule_df_list_list_preseason).items():
    print(col, ':', nb)


Nombre de valeurs null
DATE : 0
OPPONENT : 0
RESULT : 0
W-L : 0
Hi Points : 2147
Hi Rebounds : 2147
Hi Assists : 2147
Home_Away : 0
Title : 0


<span style="color:red">
Comme on peut le voir les Hi Points, Hi Rebounds, Hi Assists manquant viennent des données de présaison

In [25]:
print('nombre de classement :',len(schedule_df_list))

nombre de classement : 1678


In [26]:
print('\nNombre de valeurs == 0 ou \'\'')
for col,nb in get_empty_columns_counts(schedule_df_list).items():
    print(col, ' : ', nb)


Nombre de valeurs == 0 ou ''
DATE  :  0
OPPONENT  :  0
RESULT  :  0
W-L  :  0
Hi Points  :  0
Hi Rebounds  :  0
Hi Assists  :  0
Home_Away  :  0
Title  :  0


<span style="color:green">
Pas de columns vides ou a 0 dans tout nos classement ---> ok
</span>

### Vérification pour les données des matchs des classements (Standings)

In [27]:
total_lines = 0
for df in standing_df_list:
    total_lines += len(df)
print("Nombre d'equipe classé avec répétition :", total_lines)

print('nombre de classement:',len(schedule_df_list))

Nombre d'equipe classé avec répétition : 1316
nombre de classement: 1678


In [28]:
print('\nNombre de valeurs null')
for col,nb in get_null_counts(standing_df_list).items():
    print(col, ':', nb)


Nombre de valeurs null
Team Name : 0
W : 0
L : 0
PCT : 0
GB : 0
HOME : 0
AWAY : 0
DIV : 0
CONF : 0
PPG : 0
OPP PPG : 0
DIFF : 0
STRK : 0
L10 : 0
Title : 0


In [29]:
print('\nNombre de valeurs == 0 ou \'\'')
for col,nb in get_empty_columns_counts(standing_df_list).items():
    print(col, ' : ', nb)


Nombre de valeurs == 0 ou ''
Team Name  :  0
W  :  0
L  :  0
PCT  :  0
GB  :  0
HOME  :  0
AWAY  :  0
DIV  :  0
CONF  :  0
PPG  :  0
OPP PPG  :  0
DIFF  :  0
STRK  :  0
L10  :  0
Title  :  0


<span style="color:green">
Les données des classements sont completes
</span>

### Vérification pour les données des statistiques des equipes (team stats)

In [30]:
total_lines = 0
for df in team_stats_df_list:
    total_lines += len(df)
print("Nombre d'equipes avec répétition :", total_lines)

print('nombre de saisons pour les données des statistiques d\'equipes:',len(team_stats_df_list))

Nombre d'equipes avec répétition : 1039
nombre de saisons pour les données des statistiques d'equipes: 45


In [31]:
print('\nNombre de valeurs null pour les données de statistiques d\'equipe')
for col,nb in get_null_counts(team_stats_df_list).items():
    print(col, ':', nb)


Nombre de valeurs null pour les données de statistiques d'equipe
Rank : 0
Team Name : 0
GP : 0
PTS : 0
FGM : 0
FGA : 0
FG% : 0
3PM : 0
3PA : 0
3P% : 0
FTM : 0
FTA : 0
FT% : 0
OR : 0
DR : 0
REB : 0
AST : 0
STL : 0
BLK : 0
TO : 0
PF : 0
Title : 0


In [32]:
print('\nNombre de valeurs null pour les données de statistiques d\'equipe version 2')
for col,nb in get_null_counts(team_stats_differential_df_list).items():
    print(col, ':', nb)


Nombre de valeurs null pour les données de statistiques d'equipe version 2
Rank : 0
Team Name : 0
GP : 0
PTS : 0
FGM : 0
FGA : 0
FG% : 0
3PM : 0
3PA : 0
3P% : 0
FTM : 0
FTA : 0
FT% : 0
OR : 0
DR : 0
REB : 0
AST : 0
STL : 0
BLK : 0
TO : 0
PF : 0
Title : 0


In [33]:
print('\nNombre de valeurs null pour les données de statistiques d\'equipe version 3')
for col,nb in get_null_counts(team_stats_opponent_df_list).items():
    print(col, ':', nb)


Nombre de valeurs null pour les données de statistiques d'equipe version 3
Rank : 0
Team Name : 0
GP : 0
PTS : 0
FGM : 0
FGA : 0
FG% : 0
3PM : 0
3PA : 0
3P% : 0
FTM : 0
FTA : 0
FT% : 0
OR : 0
DR : 0
REB : 0
AST : 0
STL : 0
BLK : 0
TO : 0
PF : 0
Title : 0


In [34]:
print('\nNombre de valeurs == 0 ou \'\' pour les données de statistiques d\'equipe')
for col,nb in get_empty_columns_counts(team_stats_df_list).items():
    print(col, ' : ', nb)


Nombre de valeurs == 0 ou '' pour les données de statistiques d'equipe
Rank  :  0
Team Name  :  0
GP  :  0
PTS  :  0
FGM  :  0
FGA  :  0
FG%  :  0
3PM  :  0
3PA  :  0
3P%  :  0
FTM  :  0
FTA  :  0
FT%  :  0
OR  :  0
DR  :  0
REB  :  0
AST  :  0
STL  :  0
BLK  :  0
TO  :  0
PF  :  0
Title  :  0


In [35]:
print('\nNombre de valeurs == 0 ou \'\' pour les données de statistiques d\'equipe version 2')
for col,nb in get_empty_columns_counts(team_stats_differential_df_list).items():
    print(col, ' : ', nb)


Nombre de valeurs == 0 ou '' pour les données de statistiques d'equipe version 2
Rank  :  0
Team Name  :  0
GP  :  0
PTS  :  0
FGM  :  0
FGA  :  0
FG%  :  0
3PM  :  0
3PA  :  0
3P%  :  0
FTM  :  0
FTA  :  0
FT%  :  0
OR  :  0
DR  :  0
REB  :  0
AST  :  0
STL  :  0
BLK  :  0
TO  :  0
PF  :  0
Title  :  0


In [36]:
print('\nNombre de valeurs == 0 ou \'\' pour les données de statistiques d\'equipe version 2')
for col,nb in get_empty_columns_counts(team_stats_opponent_df_list).items():
    print(col, ' : ', nb)


Nombre de valeurs == 0 ou '' pour les données de statistiques d'equipe version 2
Rank  :  0
Team Name  :  0
GP  :  0
PTS  :  0
FGM  :  0
FGA  :  0
FG%  :  0
3PM  :  0
3PA  :  0
3P%  :  0
FTM  :  0
FTA  :  0
FT%  :  0
OR  :  0
DR  :  0
REB  :  0
AST  :  0
STL  :  0
BLK  :  0
TO  :  0
PF  :  0
Title  :  0


<span style="color:green">
Les données des statistiques d'equipes sont completes pour toutes les versions
</span>

### Résmé des vérification des valeurs : 

Pour les données de : 
- informations des joueurs : données completes
- statistiques des joueurs  : données completes à partir des années 80s (Avant les années 80s les données de 3PM,3PA,3P%,ORB,DRB,STL,BLK,TOV  ne sont pas disponible)
- matchs : manque quelques données pour Hi Points, Hi Rebounds, Hi Assists pour les matchs de Pré saison 
- classement : données completes
- statistiques (les 3 verions) : données completes